docs/project/versioning.md

```diff
@@ -0,0 +1,269 @@
+# Toolchain and language versioning
+
+<!--
+Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+Exceptions. See /LICENSE for license information.
+SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+-->
+
+[Pull request](https://github.com/carbon-language/carbon-lang/pull/4105)
+
+<!-- toc -->
+
+## Table of contents
+
+-   [Overview](#overview)
+-   [Major version increments](#major-version-increments)
+    -   [Breaking changes](#breaking-changes)
+    -   [Exclusions to what constitutes a breaking change](#exclusions-to-what-constitutes-a-breaking-change)
+-   [Minor version increments](#minor-version-increments)
+-   [Patch version increments](#patch-version-increments)
+    -   [Examples:](#examples)
+-   [Pre-release versions](#pre-release-versions)
+    -   [Release qualification pre-releases: `MAJOR.MINOR.PATCH-rc.N`](#release-qualification-pre-releases-majorminorpatch-rcn)
+    -   [Incremental development versions: `MAJOR.MINOR.PATCH-0.{nightly,dev}.N`](#incremental-development-versions-majorminorpatch-0nightlydevn)
+        -   [Nightly pre-release versions](#nightly-pre-release-versions)
+        -   [Development pre-release versions](#development-pre-release-versions)
+-   [Relevant proposal](#relevant-proposal)
+
+<!-- tocstop -->
+
+## Overview
+
+Carbon uses a single versioning scheme across both the language itself and
+toolchain, including the standard library, compiler, linker, and all development
+tools released by the main Carbon project.
+
+The scheme conforms to and is closely based on Semantic Versioning (SemVer):
+
+-   Carbon versions: `MAJOR.MINOR.PATCH`
+-   `MAJOR` increments on backwards incompatible changes, including a
+    deprecation that might trigger a build-breaking warning.
+-   `MINOR` increments only expected during early development with a major
+    version of `0`.
+-   `PATCH` increments represent bug fixes only.
+-   Pre-release suffixes:
+    -   `MAJOR.MINOR.PATCH-rc.N`: The N-th potentially viable candidate for a
+        release.
+    -   `MAJOR.MINOR.PATCH-0.nightly.YYYY.MM.DD`: A nightly incremental
+        development build on a particular day during development of that
+        version.
+    -   `MAJOR.MINOR.PATCH-0.dev`: An interactive, incremental development build
+        on a particular day by some developer during development of that
+        version.
```

Comment:
Given that we don't have `-alpha` and `-beta` here, it seems that we don't need the `-0.` prefix to ensure that `dev` and `nightly` are ordered before `rc`. Would it make sense to remove the `0.` part? (We could add it back in future if we add back `alpha` / `beta`.)

Reply:
I'm hesitant to do this... It seems nice to have `nightly` and `dev` be more explicitly separate categories from `rc` or anything else we have here.

Reply:
OK. I don't have a strong opinion. And maybe relying on the coincidence that `dev` and `nightly` collate before `rc` is too "clever".